Access control facet and lib

Rushikesh0125 opened the pull request:
Implemented access control facet and library

- exposed intuitive functions like assertOnlyRole, hasRole, and getAdminRole
- exposed functions for role management with revoke, renounce, and grant ops
- implemented a library with base internal methods
- same/similar error, event, and method names to oz

mudgen commented:
Thank you, I will be reviewing this.  We probably want to change "assert" to require because "assert" means something else in Solidity.   @0xlynett please take a look at this too.

0xlynett commented:
Also the RoleData struct feels like it could just be split into two mappings such as _roleAdmin and _hasRole

Rushikesh0125 commented:
gm, I have refactored the implementation as recommended

- renamed assert to require
- replaced getStorage references with preloaded storage ref
- removed repetitive usage of assertOwner with in-line check and removed assertOwner internal function from facet
- renamed misnamed storage position
- reorg of role data struct to two mappings

mudgen commented on src/AccessControl/libraries/LibAccessControl.sol:
```diff
@@ -0,0 +1,111 @@
+    /// @param role The role that was changed.
+    /// @param previousAdminRole The previous admin role.
+    /// @param newAdminRole The new admin role.
+    event RoleAdminChanged(bytes32 indexed role, bytes32 indexed previousAdminRole, bytes32 indexed newAdminRole);
```
All parameter names to events, errors, functions need to be prefixed with "_".  For example `_role`

panditdhamdhere reviewed (approved):
great work @Rushikesh0125
